src/cmap/worker.rs

```diff
@@ -0,0 +1,506 @@
+
+    /// Receiver for incoming pool management requests (e.g. checking in a connection).
+    management_receiver: ManagementRequestReceiver,
+
+    /// A pool manager that can be cloned and attached to connections checked out of the pool.
+    manager: PoolManager,
+}
+
+impl ConnectionPoolWorker {
+    /// Starts a worker and returns a manager and connection requester.
+    /// Once all connection requesters are dropped, the worker will stop executing
+    /// and close the pool.
+    pub(super) fn start(
+        address: StreamAddress,
+        http_client: HttpClient,
+        options: Option<ConnectionPoolOptions>,
+    ) -> (PoolManager, ConnectionRequester) {
+        let establisher = ConnectionEstablisher::new(http_client, options.as_ref());
+        let event_handler = options.as_ref().and_then(|opts| opts.event_handler.clone());
+
+        // The CMAP spec indicates that a max idle time of zero means that connections should not be
+        // closed due to idleness.
+        let mut max_idle_time = options.as_ref().and_then(|opts| opts.max_idle_time);
+        if max_idle_time == Some(Duration::from_millis(0)) {
+            max_idle_time = None;
+        }
+
+        let max_pool_size = options
+            .as_ref()
+            .and_then(|opts| opts.max_pool_size)
+            .unwrap_or(DEFAULT_MAX_POOL_SIZE);
+
+        let min_pool_size = options.as_ref().and_then(|opts| opts.min_pool_size);
+
+        let connection_options: Option<ConnectionOptions> = options
+            .as_ref()
+            .map(|pool_options| ConnectionOptions::from(pool_options.clone()));
+
+        let (handle_listener, handle) = HandleListener::new();
+        let (connection_requester, request_receiver) =
+            ConnectionRequester::new(address.clone(), handle);
+        let (manager, management_receiver) = PoolManager::new();
+
+        let worker = ConnectionPoolWorker {
+            address,
+            event_handler: event_handler.clone(),
+            max_idle_time,
+            min_pool_size,
+            establisher,
+            next_connection_id: 1,
+            total_connection_count: 0,
+            pending_connection_count: 0,
+            generation: 0,
+            connection_options,
+            available_connections: VecDeque::new(),
+            max_pool_size,
+            request_receiver,
+            management_receiver,
+            manager: manager.clone(),
+            handle_listener,
+        };
+
+        RUNTIME.execute(async move {
+            worker.execute().await;
+        });
+
+        (manager, connection_requester)
+    }
+
+    /// Run the worker thread, listening on the various receivers until all handles have been
+    /// dropped. Once all handles are dropped, the pool will close any available connections and
+    /// emit a pool closed event.
+    async fn execute(mut self) {
+        let mut maintenance_interval = RUNTIME.interval(Duration::from_millis(500));
+
+        loop {
```

Comment:
I moved the unfinished request cache to the receiver itself, which made this loop/select a little simpler.

Reply:
Awesome, this does make it a lot easier to follow!